File: src/main/java/puck/gui/imageDisplay/SVGPanel.java
package puck.gui.imageDisplay;

import org.apache.batik.dom.events.NodeEventTarget;
import org.apache.batik.dom.svg.*;
import org.apache.batik.svggen.SVGGraphics2DIOException;
import org.apache.batik.swing.JSVGCanvas;
import org.apache.batik.swing.gvt.GVTTreeRendererAdapter;
import org.apache.batik.swing.gvt.GVTTreeRendererEvent;
import org.apache.batik.swing.svg.SVGLoadEventDispatcherEvent;
import org.apache.batik.util.SVGConstants;
import org.apache.batik.util.XMLConstants;
import org.apache.batik.util.XMLResourceDescriptor;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.w3c.dom.events.Event;
import org.w3c.dom.events.EventListener;
import org.w3c.dom.events.MouseEvent;
import org.w3c.dom.svg.*;
import puck.graph.ConcreteNode;
import puck.graph.DependencyGraph;
import org.apache.batik.svggen.SVGGraphics2D;

import javax.swing.*;
import java.io.*;
import java.awt.BorderLayout;
import java.util.List;

/**
 * Created by lorilan on 3/13/15.
 */



public class SVGPanel extends JPanel{


    /*private void printDoc(){
        System.out.println(canvas.getSVGDocument().toString());
        SVGGraphics2D svgGenerator = new SVGGraphics2D(canvas.getSVGDocument());

        // Finally, stream out SVG to the standard output using UTF-8
        // character to byte encoding
        boolean useCSS = true; // we want to use CSS style attribute
        Writer out = null;
        try {
            out = new OutputStreamWriter(System.out, "UTF-8");
            svgGenerator.stream(out, useCSS);
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
        } catch (SVGGraphics2DIOException e) {
            e.printStackTrace();
        }
    }*/

    private SVGController controller;

    private DependencyGraph getGraph(){
        return controller.getGraph();
    }


    class NodeRightClickMenu extends JPopupMenu {

        //Visual actions
        JMenuItem collapse;
        JMenuItem hide;
        //Transfo actions
        JMenuItem move;

        public NodeRightClickMenu(int nodeId){
            ConcreteNode node = getGraph().getConcreteNode(nodeId);
            List<SVGController.AbstractionChoice> choices = controller.abstractionChoices(node);
            //List<JMenuItem> menuChoices = new ArrayList<>();

            JMenuItem it = new JMenuItem("Abstract " + node.name() + " as");
            it.setEnabled(false);
            this.add(it);

            for(SVGController.AbstractionChoice c : choices){
                it = new JMenuItem(c.kind() + "(" + c.policy()+ ")");
                //menuChoices.add(it);
                this.add(it);
            }

            this.addSeparator();
            hide = new JMenuItem("Hide");
            this.add(hide);
            if(getGraph().content(nodeId).nonEmpty()){
                collapse = new JMenuItem("Collapse");
                this.add(collapse);
            }

        }

    }

    class SVGPanelListener implements EventListener {

        private Integer checkIfNodeAndGetId(SVGOMTextElement txtElt){
            if(txtElt.getParentNode().getNodeName().equals("a")){
                SVGOMAElement a = (SVGOMAElement) txtElt.getParentNode();
                try{
                    return Integer.valueOf(a.getHref().getBaseVal());
                }catch(NumberFormatException e ){
                    return null;
                }
            }
            return null;
        }

        private SVGGElement getGroupingNodeParent(Node n0){

            Node n = n0.getParentNode();
            SVGGElement g;
            while(n!=null && !n.getNodeName().equals("g")){
                n = n.getParentNode();
            }
            if(n == null)
                return null;
            else {
                g = (SVGGElement) n;
                if(g.getId().startsWith("node"))
                    return g;
                else
                    return getGroupingNodeParent(g);
            }

        }

        private short LEFTBUTTON = 0;
        private short RIGHTBUTTON = 2;

       /* private Element createRectangle(float x, float y, float width, float height){

            Element fill = canvas.getSVGDocument()
                    .createElementNS(SVGDOMImplementation.SVG_NAMESPACE_URI, "rect");

            fill.setAttributeNS(null, "x", x+"");
            fill.setAttributeNS(null, "y", y+"");
            fill.setAttributeNS(null, "width", width+"");
            fill.setAttributeNS(null, "height", height+"");

            return fill;
        }

        private Element createBackgroundRectangle(SVGOMTextElement txtElt){
            SVGRect rect = txtElt.getBBox();
            Element fill = createRectangle(rect.getX(), rect.getY(), rect.getWidth(), rect.getHeight());

            fill.setAttributeNS(null, "fill", "green");
            fill.setAttributeNS(null, "stroke", "none");

            return fill;
        }*/

        private void handleLeftClick(MouseEvent evt){
            System.out.println("handling left");
            if(evt.getTarget() instanceof SVGTextElement){

                final SVGTextElement txtElt = (SVGTextElement) evt.getTarget();
                System.out.println(canvas.getUpdateManager().getClass());
                System.out.println(canvas.getUpdateManager().getUpdateRunnableQueue().getClass());

                canvas.modify(new Runnable() {
                    public void run() {
                        txtElt.setAttribute("fill", "blue"); //font color

                    }
                });
            }

            /*if(evt.getTarget() instanceof SVGPathElement){
                final SVGPathElement line = (SVGPathElement) evt.getTarget();
                System.out.println(line.getParentNode().getNodeName());
                canvas.modify(new Runnable() {
                    public void run() {
                        line.setAttribute("stroke", "blue"); //font color

                    }
                });
            }*/
        }

        private void handleRightClick(MouseEvent evt){
            if(evt.getTarget() instanceof SVGOMTextElement){
                Integer nodeId = checkIfNodeAndGetId((SVGOMTextElement) evt.getTarget());
                if(nodeId != null){
                    NodeRightClickMenu menu = new NodeRightClickMenu(nodeId);
                    menu.show(SVGPanel.this, evt.getClientX(), evt.getClientY());
                }
            }

            if(evt.getTarget() instanceof SVGOMPathElement){
                SVGOMPathElement line = (SVGOMPathElement) evt.getTarget();
                System.out.println(line.getParentNode().getNodeName());
            }
        }

        @Override
        public void handleEvent(Event evt) {
            MouseEvent mevt = (MouseEvent) evt;
            System.out.println("plop");
            if(mevt.getButton() == RIGHTBUTTON){

                System.out.println("!!!!!!!!!!!!!!!!!");
                Element elt = canvas.getElementById("clust1");

                final NodeList children = elt.getChildNodes();
                canvas.modify(new Runnable() {
                    @Override
                    public void run() {
                        for(int i=0; i< children.getLength(); i++){
                            if(children.item(i) instanceof Element) {
                                Element child = (Element) children.item(i);
                                System.out.println(child.getTagName());
                                if (child.getTagName().equals("polygon")) {
                                    child.setAttributeNS(null, "fill", "red");

                                }
                            }
                        }
                    }
                });


                handleRightClick(mevt);
            }
            if(mevt.getButton() == LEFTBUTTON){
                handleLeftClick(mevt);
            }

        }
    }


    class PUCKSVGCanvas extends JSVGCanvas {
        PUCKSVGCanvas(){
            super();
            setDocumentState(JSVGCanvas.ALWAYS_DYNAMIC);//before loading the document
            userAgent = new BridgeUserAgent(){
                @Override
                public void openLink(SVGAElement elt) {
                    //do nothing
                }
            };

            addGVTTreeRendererListener(new GVTTreeRendererAdapter() {
                @Override
                public void gvtRenderingCompleted(GVTTreeRendererEvent e) {
                    System.out.println("ready !");
                    getRoot().addEventListenerNS(XMLConstants.XML_EVENTS_NAMESPACE_URI,
                            SVGConstants.SVG_EVENT_CLICK,
                            new SVGPanelListener(),
                            false, null);
                }
            });
        }

        Element getElementById(String id){
            return getSVGDocument().getElementById(id);
        }

        void modify(Runnable r){
            getUpdateManager().getUpdateRunnableQueue().invokeLater(r);
        }
    }

    private PUCKSVGCanvas canvas;

    private NodeEventTarget getRoot(){
        return (NodeEventTarget)canvas.getSVGDocument().getRootElement();
    }

    public SVGPanel(SVGDocument doc, SVGController controller){
        canvas = new PUCKSVGCanvas();
        canvas.setSVGDocument(doc);

        this.controller = controller;
        this.setLayout(new BorderLayout());
        this.add("Center", canvas);


    }

    public static SVGPanel fromStream(InputStream stream, SVGController g) throws IOException {
        String parser = XMLResourceDescriptor.getXMLParserClassName();
        SAXSVGDocumentFactory factory = new SAXSVGDocumentFactory(parser);
        SVGDocument doc = factory.createSVGDocument("", stream);
        return new SVGPanel(doc, g);
    }

   /* public static SVGPanel fromFilePath(String filePath ) throws IOException {
        return fromStream(new FileReader(filePath));
    }*/
}

File: puckGui/src/main/scala/puck/gui/svg/SVGPanel.scala
package puck.gui.svg

import java.awt.BorderLayout
import java.awt.event.{InputEvent, KeyEvent}
import java.util.regex.{Matcher, Pattern}
import javax.swing.{JPopupMenu, KeyStroke, JPanel}

import org.apache.batik.dom.GenericText
import org.apache.batik.dom.events.NodeEventTarget
import org.apache.batik.swing.JSVGCanvas
import org.apache.batik.swing.gvt.{GVTTreeRendererEvent, GVTTreeRendererAdapter}
import org.apache.batik.swing.svg.AbstractJSVGComponent
import org.apache.batik.util.{SVGConstants, XMLConstants}
import org.w3c.dom.{Node, NodeList, Element}
import org.w3c.dom.events.{EventListener, Event, MouseEvent}
import org.w3c.dom.svg._
import puck.graph.{Isa, Uses, NodeId, NodeIdP, DGEdge}
import puck.gui.svg.SVGPanel._


class PUCKSVGCanvas
( panel : SVGPanel,
  eventListenerBuilder : EventListenerBuilder
  ) extends JSVGCanvas {
  setDocumentState(AbstractJSVGComponent.ALWAYS_DYNAMIC)

  userAgent = new BridgeUserAgent(){
    override def openLink(elt: SVGAElement) : Unit = ()
  }

  def root = getSVGDocument.getRootElement.asInstanceOf[NodeEventTarget]

  import KeyEvent.{VK_EQUALS, VK_ADD, VK_MINUS, VK_SUBTRACT}
  import InputEvent.{CTRL_DOWN_MASK, CTRL_MASK, SHIFT_DOWN_MASK}
  import JSVGCanvas.{ZOOM_IN_ACTION, ZOOM_OUT_ACTION}
  import KeyStroke.getKeyStroke

  val ctrl_add = getKeyStroke(VK_EQUALS, CTRL_DOWN_MASK | SHIFT_DOWN_MASK)
  getInputMap.put(ctrl_add, ZOOM_IN_ACTION)
  val ctrl_add_numpad = getKeyStroke (VK_ADD, CTRL_DOWN_MASK)
  getInputMap.put (ctrl_add_numpad, ZOOM_IN_ACTION)
  val ctrl_minus = getKeyStroke (VK_MINUS, CTRL_MASK)
  getInputMap.put (ctrl_minus, ZOOM_OUT_ACTION)
  val ctrl_minus_numpad = getKeyStroke (VK_SUBTRACT, CTRL_MASK)
  getInputMap.put (ctrl_minus_numpad, ZOOM_OUT_ACTION)

  addGVTTreeRendererListener (new GVTTreeRendererAdapter() {
    override def gvtRenderingCompleted (e: GVTTreeRendererEvent) : Unit = {
      root.addEventListenerNS(XMLConstants.XML_EVENTS_NAMESPACE_URI,
        SVGConstants.SVG_EVENT_CLICK,
        eventListenerBuilder(PUCKSVGCanvas.this, panel), false, null)
    }
  })

  def getElementById (id: String): Element =  getSVGDocument getElementById id

  def modify (f : () => Unit ) : Unit =
    getUpdateManager.getUpdateRunnableQueue.invokeLater(new Runnable {
      def run(): Unit = f()
    })

}

class SVGPanelListener
( canvas : PUCKSVGCanvas,
  panel : SVGPanel)
  extends EventListener {

  def controller = panel.controller

  private def checkIfNodeAndGetId(txtElt: Element): Option[Int] = {
    if (txtElt.getParentNode.getNodeName == "a") {
      val a: SVGAElement = txtElt.getParentNode.asInstanceOf[SVGAElement]
      try Some(a.getHref.getBaseVal.toInt)
      catch {
        case e: NumberFormatException => None
      }
    } else None
  }

  private def checkIfEdgeAndGetGElement(path: Element): Option[SVGGElement] = {
    if (path.getParentNode.getNodeName == "g") {
      val gelt: SVGGElement = path.getParentNode.asInstanceOf[SVGGElement]
      if (gelt.getId.startsWith("edge")) Some(gelt)
      else None
    } else None
  }

  private val idPattern = Pattern.compile("(\\d+:\\d+)|(nodeCluster\\d+)")

  private def edgeFromGElement(gelt: SVGGElement): Option[NodeIdP] = {
    val t = gelt.getFirstChild.asInstanceOf[SVGTitleElement]
    val title = t.getFirstChild.asInstanceOf[GenericText]

    val m: Matcher = idPattern.matcher(title.getData)

    def nextId() : Option[Int] = {
      if(m.find()) {
        val strId = m.group()
        if(strId.startsWith("nodeCluster")){
          Some(strId.substring("nodeCluster".length).toInt)
        } else {
          Some(strId.substring(strId.indexOf(":") + 1).toInt)
        }
      } else None
    }
    val ssrc = nextId()
    val stgt = nextId()
    (ssrc, stgt) match {
      case (Some(source), Some(target)) => Some((source, target))
      case _ => None
    }
  }

  private def changeEdgeColor(edge: SVGGElement, color: String) : Unit = {
    val l: NodeList = edge.getChildNodes

    var i: Int = 0
    while (i < l.getLength) {
      val n: Node = l.item(i)
      if (n.getNodeName == "path") {
        n.asInstanceOf[Element].setAttribute("stroke", color)
      }
      if (n.getNodeName == "polygon") {
        n.asInstanceOf[Element].setAttribute("stroke", color)
        n.asInstanceOf[Element].setAttribute("fill", color)
      }
      i += 1
    }
  }

  val setNodeColor : ((NodeId, String, Element)) => Unit = {
    case ((_, color, domElt)) =>
      domElt.setAttribute("fill", color)
  }
  private def resetAllNodes() : Unit = {
    if(controller.selectedNodes.nonEmpty){
      controller.selectedNodes foreach setNodeColor
      controller.resetSelectedNodes()
    }
  }
  private def conditionalEdgeReset(): Option[NodeIdP] = {
    controller.selectedEdge match {
      case Some((e, color, domElt)) =>
        changeEdgeColor(domElt, color)
        controller.resetEdgeSelected()
        Some(e)
      case None => None
    }
  }

  private val selectColor: String = "blue"
  private val LEFTBUTTON: Short = 0
  private val RIGHTBUTTON: Short = 2

  private def handleLeftClick(evt: MouseEvent) : Unit = {
      val shiftPressed : Boolean = evt.getShiftKey
      evt.getTarget match {
        case txtElt: SVGTextElement =>
          checkIfNodeAndGetId(txtElt).foreach {
            nodeClickedId =>
              canvas.modify { () =>
                conditionalEdgeReset()

                if(!shiftPressed) {
                  val removed = controller.keepOnlySelectedNode(nodeClickedId)
                  removed foreach setNodeColor
                }
                controller.getSelectedNode(nodeClickedId) match {
                  case Some((_, color, domElt)) =>
                    controller.removeSelectedNode(nodeClickedId)
                    domElt.setAttribute("fill", color)
                  case None =>
                    controller.addNodeToSelection(nodeClickedId, txtElt)
                    txtElt.setAttribute("fill", selectColor)
                }
              }
          }
        case _: SVGPathElement
             | _: SVGPolygonElement =>
          val line: Element = evt.getTarget.asInstanceOf[Element]

          canvas.modify { () =>
            if(controller.selectedNodes.nonEmpty)
              controller.resetSelectedNodes()

            val sPrevEdge = conditionalEdgeReset()
            checkIfEdgeAndGetGElement(line) foreach {
              gedge =>
                val sEdge = edgeFromGElement(gedge)

                sEdge foreach {
                  e =>
                    if (sEdge != sPrevEdge) {
                      val color: String = line.getAttribute("stroke")
                      controller.setEdgeSelected(e, gedge, color)
                      changeEdgeColor(gedge, selectColor)

                    }
                }
            }
          }
        case _ =>
          println("reset both")
          canvas.modify { () =>
            if(controller.selectedNodes.nonEmpty)
              controller.resetSelectedNodes()
            conditionalEdgeReset()
            ()
          }
      }
    }

  private def handleRightClick(evt: MouseEvent) : Unit =
     evt.getTarget match {
        case txtElt: SVGTextElement =>
          checkIfNodeAndGetId(txtElt) foreach {
            nodeId =>
              val menu: JPopupMenu = NodeRightClickMenu(controller, nodeId)
              menu.show(panel, evt.getClientX, evt.getClientY)
          }
        case _: SVGPathElement
             | _: SVGPolygonElement =>
          val line: Element = evt.getTarget.asInstanceOf[Element]
          checkIfEdgeAndGetGElement(line) foreach {
            gedge =>
              edgeFromGElement(gedge) foreach {
                e =>
                  val menu: EdgeRightClickMenu = new EdgeRightClickMenu(controller, e)
                  menu.show(panel, evt.getClientX, evt.getClientY)
              }
          }
      }


  def handleEvent(evt: Event) : Unit = {
    val mevt: MouseEvent = evt.asInstanceOf[MouseEvent]
    mevt.getButton match {
      case RIGHTBUTTON => handleRightClick(mevt)
      case LEFTBUTTON => handleLeftClick(mevt)
      case _ => Console.err.println("mouse button event unknown")
    }
  }
}

object SVGPanel {
  type EventListenerBuilder = (PUCKSVGCanvas, SVGPanel) => EventListener
  val defaultListener : EventListenerBuilder =
    (canvas, panel) => new SVGPanelListener(canvas, panel)
  val deafListener : EventListenerBuilder =
    (_,_) => new EventListener {
      override def handleEvent(evt: Event): Unit = ()
    }
}

class SVGPanel
( doc : SVGDocument,
  eventListenerBuilder: EventListenerBuilder = defaultListener) extends JPanel {
  val canvas = new PUCKSVGCanvas(this, eventListenerBuilder)
  canvas.setSVGDocument(doc)
  setLayout(new BorderLayout())
  add("Center", canvas)

  var controller : SVGController = _

}
